src/Symfony/Component/String/Tests/AbstractAsciiTestCase.php

```diff
@@ -0,0 +1,730 @@
+    }
+
+    public static function provideLower()
+    {
+        return [
+            ['hello world', 'hello world'],
+            ['hello world', 'HELLO WORLD'],
+            ['hello world', 'Hello World'],
+            ['symfony', 'symfony'],
+            ['symfony', 'Symfony'],
+            ['symfony', 'sYmFOny'],
+        ];
+    }
+
+    /**
+     * @dataProvider provideUpper
+     */
+    public function testUpper(string $expected, string $origin)
+    {
+        $instance = static::createFromString($origin)->upper();
+
+        $this->assertNotSame(static::createFromString($origin), $instance);
+        $this->assertEquals(static::createFromString($expected), $instance);
+        $this->assertSame($expected, (string) $instance);
+    }
+
+    public static function provideUpper()
+    {
+        return [
+            ['HELLO WORLD', 'hello world'],
+            ['HELLO WORLD', 'HELLO WORLD'],
+            ['HELLO WORLD', 'Hello World'],
+            ['SYMFONY', 'symfony'],
+            ['SYMFONY', 'Symfony'],
+            ['SYMFONY', 'sYmFOny'],
+        ];
+    }
+
+    /**
+     * @dataProvider provideTitle
+     */
+    public function testTitle(string $expected, string $origin, bool $allWords)
+    {
+        $this->assertEquals(
+            static::createFromString($expected),
+            static::createFromString($origin)->title($allWords)
+        );
+    }
+
+    public static function provideTitle()
+    {
+        return [
+            ['Hello world', 'hello world', false],
+            ['Hello World', 'hello world', true],
+            ['HELLO WORLD', 'HELLO WORLD', false],
+            ['HELLO WORLD', 'HELLO WORLD', true],
+            ['HELLO wORLD', 'hELLO wORLD', false],
+            ['HELLO WORLD', 'hELLO wORLD', true],
+            ['Symfony', 'symfony', false],
+            ['Symfony', 'Symfony', false],
+            ['SYmFOny', 'sYmFOny', false],
+        ];
+    }
+
+    /**
+     * @dataProvider provideSlice
+     */
+    public function testSlice(string $expected, string $origin, int $start, int $length = null)
+    {
+        $this->assertEquals(
+            static::createFromString($expected),
+            static::createFromString($origin)->slice($start, $length)
+        );
+    }
+
+    public static function provideSlice()
+    {
+        return [
+            ['Symfony', 'Symfony is awesome', 0, 7],
+            [' ', 'Symfony is awesome', 7, 1],
+            ['is', 'Symfony is awesome', 8, 2],
+            [' ', 'Symfony is awesome', 10, 1],
+            ['awesome', 'Symfony is awesome', 11, 7],
+        ];
+    }
+
+    /**
+     * @dataProvider provideAppend
+     */
+    public function testAppend(string $expected, array $suffixes)
+    {
+        $instance = static::createFromString('');
+        foreach ($suffixes as $suffix) {
+            $instance = $instance->append($suffix);
+        }
+
+        $this->assertEquals(static::createFromString($expected), $instance);
```

Comment:
Maybe extending the test and do also 

$instance = $instance->append(...$suffix);

and then check the expected..

On the same issue I do wonder if having variadic on append/prepend might become confusing...

Reply:
good idea, updated.
about variadics, I added it because ppl requested it.